### PR TITLE
UGDG Update

### PR DESCRIPTION
**Pull request by krusagiz**
Updated UG for Tag/Untag and Convert commands

Updated DG for Tag/Untag commands

### CODE REVIEW

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+
+The **tag**/**untag** command performs modifications on existing `**Transaction**` entries. Due to polymorphism, both **Expense**` and `**Income**` objects extends the abstract class `**Transaction**`, and are treated the same way in the context of the **tag**/**untag** command.
+
+.A Class Diagram illustrating the implementation of Tags inside a Transaction
+image::tag-untag/Polymorphism.png[]
+
+Because of polymorphism, many of the driver functions in `THRIFT` uses `Transaction` for both `Expense`
+and `Income` objects. For example, when tagging a new `Expense`/`Income`, the `TagCommandParser` will treat them as the same object:
+
+image::tag-untag/TagSequence.png[]
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc:294]: File seem to be missing

File: docs/DeveloperGuide.adoc [hunk 2]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+. Incorrect syntax (i.e. missing prefix, if it is required)
+. Illegal values in parameters (i.e. special characters and symbols entered for an integer-only field)
+. Multiple occurrences of parameters which only expects a single entry
+
+[NOTE]
+If the user input is incorrect due to any of the reasons above, the corresponding usage syntax will be shown.
+
+The following is an example on how a `**Tag**` is added and removed from a `**Transaction**`, with details on the processes in the backend.
+
+Step 1. The user launches the application with data from previous sessions. THRIFT does contains only one `**Transaction**` containing no tags and the user is going to perform tagging and un-tagging operations on it.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: `**THRIFT**` ~~does~~ contains

File: docs/DeveloperGuide.adoc [hunk 3]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+
+image::tag-untag/TagUntagTransactionWoTag.png[]
+
+Step 2. The user executes the command **tag i/1 t/Delicious** to tag the `**Transaction**` at `**Index**` 1 with the `**Tag**` "Delicious". The input is checked with a `**TagCommandParser**` parser and an attempt to parse each parameter occurs:
+
+* `**Index**` is parsed by `ParserUtil#parseIndex(String)`
+* `**Tag**` strings are parsed by `ParserUtil#parseTags(Collection<String>)` which calls `ParserUtil#parseTag(String)` iteratively for every string in the collection
+
+TIP: `**ParserUtil**` is a class that contains useful functions for parsing the inputs from the user
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Should add full-stop at the end since it sounds like a complete sentence.

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+
+[NOTE]
+Tags that already exist inside the specified `**Transaction**` will be ignored, and if that results in no tags being updated, an error will be shown to the user.
+
+Since the user input is valid, a `**TagCommand**` command object is created and executed. As a result, a copy of the `**Transaction**` object with the new set of `**Tag**` objects with the specified `**Tag**` inside replaces the original in the `**TransactionList**` at the same position.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: The second sentence sounds weirdly long. Try fragmenting it?

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+
+image::tag-untag/TagUntagTransactionWTag.png[]
+
+Step 3. The user realises that the Tag added was not appropriate. The user then executes the command **untag i/1 t/Delicious** to untag the `**Tag**` "Delicious" from the `**Transaction**` at `**Index**` 1. The input is now checked with a `**UntagCommandParser**` parser and similarly, an attempt to parse each parameter occurs.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: `**Tag**`

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+
+* `**Index**` is parsed by `ParserUtil#parseIndex(String)`
+* `**Tag**` strings are parsed by `ParserUtil#parseTags(Collection<String>)` which calls `ParserUtil#parseTag(String)` iteratively for every string in the collection
+
+[NOTE]
+Tags that do not exist inside the specified `**Transaction**` will be ignored, and if that results in no tags being updated, an error will be shown to the user.
+
+Again, since the user input is valid, a `**UntagCommand**` command object is created and executed. As a result, a copy of the `**Transaction**` object with the new set of `**Tag**` objects with the specified `**Tag**` removed replaces the original in the `**TransactionList**` at the same position.
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: Sentence structure - too long

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -278,7 +278,88 @@ are applied when restoring `**THRIFT**` data from the data file.
...
+image::tag-untag/TagUntagTransactionWoTag.png[]
+
+//The following activity diagram summarises what happens the user executes a command to **tag** a `**Transaction**`. The **untag** command follows the same sequence.
+
+
+==== Design Considerations
+
+===== Aspect: Mutability of `**Transaction**` objects
+
+* **Alternative 1: (current choice)**: Create a copy of the `**Transaction**` object with modified tags, and replace the original in the `**TransactionList**`
```
Comment (limerencee) [inline docs/DeveloperGuide.adoc]: ... object with modified tags and replace ...

**Review by limerencee — CHANGES_REQUESTED** [review]
LGTM after some minor corrections!

File: docs/DeveloperGuide.adoc (hunk 1 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc:294]: Forgot to add it to staging, now fixed.

File: docs/DeveloperGuide.adoc (hunk 3 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Full-stop has been added.

File: docs/DeveloperGuide.adoc (hunk 2 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Corrected Grandma.

File: docs/DeveloperGuide.adoc (hunk 4 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: I removed some parts.

File: docs/DeveloperGuide.adoc (hunk 6 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: I removed some parts of the sentence.

File: docs/DeveloperGuide.adoc (hunk 7 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Removed the comma.

File: docs/DeveloperGuide.adoc (hunk 5 above)
Comment (krusagiz) [reply docs/DeveloperGuide.adoc]: Added the bold.

**Review by limerencee — APPROVED** [review]

**Pull request merged**